### PR TITLE
Moves pysmurf-controller functions that load lots of data to subprocess

### PR DESCRIPTION
**Pull request by jlashner**
This PR changes pysmurf-controller operations that load TODs such that instead of running in the main process, they call out to a subprocess.

## Description

This introduces the `smurf_subprocess_util` module, and a protocol for running functions inside a twisted subprocess in such a way that a general RunCfg can be passed to it, and a general RunResult can be returned. Any operation that involves TOD data loading is modified so that the loading and analysis occurs in a subprocess rather than in the main one.

## Motivation and Context

This should hopefully solve the memory leak, which I believe is coming from instances where TOD data is not being released to the system after it leaves scope. Moving TOD loading and analysis into subprocesses instead of the main one should enforce that this memory is returned to the OS when the subprocess exits. [This discussion](https://github.com/simonsobs/daq-discussions/discussions/81#discussioncomment-9286023) has some links that motivate this solution.

## How Has This Been Tested?

I have tested the subprocess protocol locally through the use of the `test` process (and the new `run_test_func` operation).
I will need to test all other operations with hardware on SATp1 or SATp3.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: socs/agents/pysmurf_controller/smurf_subprocess_util.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    Args
+    -----
+    cfg: RunCfg
+        Configuration object to specify the function to run, and the arguments.
+    """
+    if not in_reactor_context():
+        return threads.blockingCallFromThread(
+            reactor, _run_func_in_subprocess_reactor, cfg)
+    else:
+        return _run_func_in_subprocess_reactor(cfg)
```
Comment (mhasself) [inline socs/agents/pysmurf_controller/smurf_subprocess_util.py:264]: Cool... I believe that if you _are_ in reactor context, then you can't block -- i.e you need to return a Deferred here, which would only be possible if `_run_func_in_subprocess_reactor(cfg)` returned `prot.deferred` or some wrapped thing that applies the Runresult.

Comment (jlashner) [reply socs/agents/pysmurf_controller/smurf_subprocess_util.py]: Good point... I changed to `run_func_in_subprocess_from_thread` and had it only handle the from-thread case

**Comment by jlashner** [comment]
This has been tested on satp1 and it works! It seems to have fixed the large memory leak. Notes from my testing is here: https://simonsobs.atlassian.net/wiki/spaces/~55705801070cc775254537ab15f2dcf6702b78/pages/412778522/2024-05-02+pysmurf-controller+subprocessing+tests

Notably, memory returns to base-level after IVs and bias steps, which it did not previously: 
image

I want to note that there does still seem to be a memory leak caused by the functions that have _not_ been moved to subprocesses. I did not move the `stream` data function to a subprocess because it isn't loading TODs, however just the creation of the pysmurf object seems to be leaking memory, at a rate of ~15 MB per call:
image

I think we should probably move _all logic_ involving pysmurf to a subprocess, however I this this PR solves the most pressing issues and largest memory leaks, so I think we should merge this first.

**Review by BrianJKoopman — CHANGES_REQUESTED** [review]
Glad the subprocess route is working! Can you address the unused imports caught by pre-commit: https://results.pre-commit.ci/run/github/186511668/1714677193.715HK9dWRnKVTIpCyOTUcg

File: socs/agents/pysmurf_controller/smurf_subprocess_util.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    }
+    quantiles = np.array([15, 25, 50, 75, 85])
+    data['quantiles'] = _process_quantiles(quantiles, arrays)
+    return data
+
+
+def test():
+    """Function for testing subprocess operation"""
+    print("HEREE")
+    return {'test': 10}
```
Comment (BrianJKoopman) [inline socs/agents/pysmurf_controller/smurf_subprocess_util.py:193]: Remove? Especially if subprocess task dropped from Agent code.

File: socs/agents/pysmurf_controller/smurf_subprocess_util.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    def processExited(self, status):
+        if self.result is None:
+            raise RuntimeError("No result received from child process")
+
+        self.deferred.callback(self.result)
+
+
+@defer.inlineCallbacks
+def _run_func_in_subprocess_reactor(cfg: RunCfg) -> RunResult:
+    """Helper function for run_func_in_subprocess, that can assume reactor context"""
```
Comment (BrianJKoopman) [inline socs/agents/pysmurf_controller/smurf_subprocess_util.py:270]: Is `run_func_in_subprocess` now `run_smurf_func`? If not, where is this function?

**Review by BrianJKoopman — COMMENTED** [review]
This looks good to me, a few small comments/questions below.

File: socs/agents/pysmurf_controller/agent.py [hunk 4]
```diff
@@ -332,6 +342,24 @@ def _stop_check_state(self, session, params):
...
+        if not result.success:
+            self.log.error("Subprocess errored out:\n{tb}", tb=result.traceback)
+
+        if isinstance(result.return_val, dict):
+            session.data = result.return_val
+        else:
+            session.data['data'] = result.return_val
+
+        return result.success, 'finished'
+
```
[the remark this answers is not in the document]
Comment (jlashner) [reply socs/agents/pysmurf_controller/agent.py:362]: I do want to keep this since it is helpful for testing locally, but I will correct docstrings.

**Review by BrianJKoopman — COMMENTED** [review]
Addressed comments look good, merge conflicts still need resolving though.

**Review by BrianJKoopman — APPROVED** [review]

**Pull request merged**